lib/storage/metadata/oplog/BucketdOplogInterface.js

```diff
@@ -0,0 +1,257 @@
+                    }
+                    this.persist.save(
+                        filterName, this.persistData, cseq, err => {
+                            if (err) {
+                                return next(err);
+                            }
+                            return next(null, raftId, cseq);
+                        });
+                    return undefined;
+                });
+                return undefined;
+            },
+            /*
+             * In this step we loop over the oplog
+             */
+            (raftId, cseq, next) => {
+                this.logger.info(`reading oplog raftId=${raftId} cseq=${cseq}`,
+                                 { filterName });
+                // only way to get out of the loop in all cases
+                const nextOnce = jsutil.once(next);
+                let doStop = false;
+                // resume reading the oplog from cseq. changes are idempotent
+                const logConsumer = new LogConsumer({
+                    bucketClient: this.bkClient,
+                    raftSession: raftId,
+                });
+                let _cseq = cseq;
+                async.until(
+                    () => doStop,
+                    _next => {
+                        logConsumer.readRecords({
+                            startSeq: _cseq,
+                            limit: this.bucketdOplogQuerySize,
+                        }, (err, record) => {
+                            if (err) {
+                                this.logger.error('readRecords error', { err, filterName });
+                                return setTimeout(() => _next(), 5000);
+                            }
+                            if (!record.log) {
+                                // nothing to read
+                                return setTimeout(() => _next(), 5000);
+                            }
+                            const seqs = [];
+                            record.log.on('data', chunk => {
+                                seqs.push(chunk);
+                            });
+                            record.log.on('end', () => {
+                                const addQueue = [];
+                                const delQueue = [];
+                                for (let i = 0; i < seqs.length; i++) {
+                                    if (filter.filterType === 'raftSession' ||
+                                        (filter.filterType === 'bucket' &&
+                                         seqs[i].db === filter.bucket.bucketName)) {
+                                        for (let j = 0; j < seqs[i].entries.length; j++) {
+                                            const _item = {};
+                                            _item.bucketName = seqs[i].db;
+                                            _item.key = seqs[i].entries[j].key;
+                                            if (seqs[i].entries[j].type !== undefined &&
```

Comment:
This block is tabbed over quite a bit, could we break this out or maybe use some guard clauses here? I can't even tell what is going on anymore.